posix/device.c

```diff
@@ -0,0 +1,321 @@
+
+/*
+**==============================================================================
+**
+** Public functions:
+**
+**==============================================================================
+*/
+
+int oe_device_table_set(uint64_t devid, oe_device_t* device)
+{
+    int ret = -1;
+    bool locked = false;
+
+    if (!device)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    oe_conditional_lock(&_lock, &locked);
+
+#if !defined(NDEBUG)
+    _assert_device(device);
+#endif
+
+    if (_resize_table(devid + 1) != 0)
+        OE_RAISE_ERRNO(OE_ENOMEM);
+
+    if (_table[devid] != NULL)
+        OE_RAISE_ERRNO(OE_EEXIST);
+
+    _table[devid] = device;
+
+    ret = 0;
+
+done:
+    oe_conditional_unlock(&_lock, &locked);
+
+    return ret;
+}
+
+static oe_device_t* _get_device(uint64_t devid, oe_device_type_t type)
+{
+    oe_device_t* ret = NULL;
+    oe_device_t* device;
+
+    if (devid >= _table_size)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    device = _table[devid];
+
+    if (device && type != OE_DEVICE_TYPE_ANY && device->type != type)
+        goto done;
+
+    ret = device;
+
+done:
+
+    return ret;
+}
+
+oe_device_t* oe_device_table_get(uint64_t devid, oe_device_type_t type)
+{
+    oe_device_t* ret;
+
+    oe_conditional_lock(&_lock, NULL);
+    ret = _get_device(devid, type);
+    oe_conditional_unlock(&_lock, NULL);
+
+    return ret;
+}
+
+oe_device_t* oe_device_table_find(const char* name, oe_device_type_t type)
+{
+    oe_device_t* ret = NULL;
+    oe_device_t* device = NULL;
+    size_t i;
+    bool locked = false;
+
+    if (!name)
+        goto done;
+
+    oe_conditional_lock(&_lock, &locked);
+
+    for (i = 0; i < _table_size; i++)
+    {
+        oe_device_t* p = _table[i];
+
+        if (p && oe_strcmp(p->name, name) == 0)
+        {
+            device = p;
+            break;
+        }
+    }
+
+    if (device && type != OE_DEVICE_TYPE_ANY && device->type != type)
+        goto done;
+
+    ret = device;
+
+done:
+    oe_conditional_unlock(&_lock, &locked);
+
+    return ret;
+}
+
+int oe_device_table_remove(uint64_t devid)
+{
+    int ret = -1;
+    oe_device_t* device;
+    bool locked = false;
+
+    oe_conditional_lock(&_lock, &locked);
+
+    if (!(device = _get_device(devid, OE_DEVICE_TYPE_ANY)))
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    if (devid >= _table_size || _table[devid] == NULL)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    _table[devid] = NULL;
+
+    if (device->ops.device.release(device) != 0)
+        OE_RAISE_ERRNO(oe_errno);
+
+    ret = 0;
+
+done:
+    oe_conditional_unlock(&_lock, &locked);
+
+    return ret;
+}
+
+/*
+**==============================================================================
+**
+** oe_set_thread_devid()
+** oe_get_thread_devid()
+** oe_clear_thread_devid()
+**
+**==============================================================================
+*/
+
+static oe_once_t _tls_device_once = OE_ONCE_INIT;
+static oe_thread_key_t _tls_device_key = OE_THREADKEY_INITIALIZER;
```

Comment:
Most of this manual thread-specific management can be replaced by using the GNU __thread keyword which would lead to an optimized and cleaner thread-local implementation

Reply:
Fixed. Yes, much cleaner now. :)